### DISCUSSION TITLE
bytes per row slider

### DISCUSSION THREAD

**Discussion by QuantumApprentice** [WerWolv/ImHex#1107] [Feedback]
I have a workflow that requires inspecting small images of different widths, and I've found it very helpful to visually see the image frames by adjusting the bytes per row to match their widths.

Unfortunately the slider for this is hidden in the "Extras->Settings" menu and doesn't have a lot of direct adjust-ability, but fortunately I can use ctrl-click to manually set the value.

It would be nice to have this on the front page (probably above the address bar) and expand the range to something in the 1k to 5k range, or at least prevent the program from crashing if an excessively large value is set, or maybe allow the user to set the upper range value.

**Comment by WerWolv** [comment]
1k to 5k is an absolute MASSIVE number of bytes that need to be displayed on the screen and this is in my opinion really not how this feature is intended to be used.

Instead you should consider writing a pattern that highlights the images (and potentially use the bitmap visualizer to turn it into an image right away)

**Reply by shodanx2** [reply]
Hello,
First time using imhex today.
How can I view more bits is my first question that lead me here.
I don't understand why the hex editor view is not set to auto-resize with its view
Here is what it looks like to me right now. Only 137 rows of 16 bytes each.

![image](https://github.com/WerWolv/ImHex/assets/10621885/3fa5b533-701a-41ec-b67f-375cdc9cb3d6)

On my 4k  40" TV there is space to show 112 by 137, or 15K of data

![image](https://github.com/WerWolv/ImHex/assets/10621885/b1225215-dbc8-4e93-89d6-5d3a28d66adb)

And that's if I don't use my side-monitors

In which case 45K of data can be seen at once

![image](https://github.com/WerWolv/ImHex/assets/10621885/52ae8447-da2c-47e5-9c26-249c8d7b7b6b)

**Reply by WerWolv** [reply]
You can click on the little up arrow in the hex editor footer and then adjust the value of the marked slider. By default it only goes up to 32 but you can CTRL click it and enter a much larger number if you want to

![image](https://github.com/WerWolv/ImHex/assets/10835354/80220aa1-3f14-4614-9b20-d6673a368872)

About your proposal, keep in mind that ImHex uses Immediate Mode rendering. That means the entire data you're seeing on the screen is being re-rendered every single frame. So the thing you want will require a pretty beefy GPU to work nicely. Besides, personally I think having this much data on the screen will just give you cognitive overload, at least it would do that to me.

**Comment by QuantumApprentice** [comment]
"1k to 5k is an absolute MASSIVE number of bytes that need to be displayed on the screen"
1) I don't need them all displayed on the screen at the same time.
2) Microsoft Visual Studio can show 512, and HxD Hex Editor (my go to hex editor until I was introduced to this one) will go as high as 4096, 1k-5k is not unusual.
The bigger problem for me is this editor crashes whenever I put anything above 108 into the slider.

"this is in my opinion really not how this feature is intended to be used."
3) Interesting opinion, not sure how most people think a hex editor should be used, but ok.
I'm reverse engineering a proprietary image format from a game from 1996.

"Instead you should consider writing a pattern that highlights the images"
4) This sounds useful, I'll look at the pattern language, but at a brief glance I'm not sure I can use this to align the image in such a way to make it viewable because...

"(and potentially use the bitmap visualizer to turn it into an image right away)"
5) The image format is 8-bit paletted indexed, based on a 256 color palette, some of which cycle.
Ignoring the cycling, the indexing itself doesn't translate easily to a 3 color bitmap, but I can easily see the outlines of the image itself (and thus where a frame [there are multiple frames in a file] begins and ends) if I just set the slider width to match the image width.

That being said, the visualizer looks cool, but I can't figure out how to get it to visualize anything.
Is there a tutorial I can follow that shows how to get the Read node to access the data starting at a hex address? I keep putting the address in but it keeps removing the hex value (maybe truncating it to decimal? the docs don't say). Selected region doesn't seem to work as described either.

I'm sure I'm just doing something wrong, but I can't figure out what right away :(

Whatever you decide to do with this feedback, I like this hex editor.
It seems to have a bunch of interesting features that I'm interested in, just wish I had a better understanding how to use.
Nice job! :)

**Reply by Calinou** [reply]
> 5) The image format is 8-bit paletted indexed, based on a 256 color palette, some of which cycle.
Ignoring the cycling, the indexing itself doesn't translate easily to a 3 color bitmap, but I can easily see the outlines of the image itself (and thus where a frame [there are multiple frames in a file] begins and ends) if I just set the slider width to match the image width.

I wonder if it's possible to get the correct palette to show up in the image visualizer using the [Data Processor](https://docs.werwolv.net/imhex/views/data-processor) to convert the paletted color data to RGBA8 colors.

That said, this would require porting the palette to the node setup, or reading it from the binary if it's stored there.

**Comment by WerWolv** [comment]
A lot of things have changed since. There's now a little popup menu in the footer of the hex editor view that lets you change various settings about how things in the Hex Editor should be rendered. Besides the number of columns, it also allows you to change the type of value that's displayed in the cells, if you want a separator line every N bytes or if addresses and co should be displayed in hex, decimal or octal :)

<img width="276" alt="image" src="https://github.com/user-attachments/assets/d5f347e2-ad2e-48ee-a482-e19a2192b39d" />

**Reply by shodanx2** [reply]
Thank you for notifying me of this change !
I was using ImHex to view El Torito boot fields of ISO files last time I needed ImHex.

Ultimately I could do the job with 16 columns but it would have been easier to have a greater view.

I have just tested these changes and this is already a great improvement. 

The performance is unchanged even with the maximum of 128 columns enabled.

However now I see I could probably do 500 columns just fine.

It is difficult to depict what this looks like, here is a video screenshot but it might not display properly for you without a display of resolution of 11520x2160

Even VLC refuses to play a file with such a resolution.

Let's see if github has been more forward thinking.


https://github.com/user-attachments/assets/4fc29ea6-f05b-471d-83d0-b1cd9cb76435

Handbrake could not use encore slower than M8 for resolution above 8K so this is probably a video of low quality despite 1900kbps AV-1

I am surprised, github could play this back so well !
Only way to watch it in my resolution is enabling "picture-in-picture" mode and then stretching the window across all monitors. Then the text becomes easily readable.

**Reply by heeplr** [reply]
> There's now a little popup menu in the footer of the hex editor view that lets you change various settings about how things in the Hex Editor should be rendered. 

This is an awesome feature but the slider doesn't really work for me for very small numbers. Is there a way to enter numbers with the keyboard?

**Comment by paxcut** [comment]
> This is an awesome feature but the slider doesn't really work for me for very small numbers. Is there a way to enter numbers with the keyboard?

All sliders can be ctrl-clicked to enable keyboard input. Note that this bypasses the bounds checks that are inherent in the sliding mechanic so it is possible to enter invalid values that may unwanted results.

**Comment by shodanx2** [comment]
Thanks !

I managed to increase the columns count to 255 before something broke 

Here is what that looks like

![2025-06-07_16h14s03_imhex-gui_ImHex_-_win10 iso_11520x2110](https://github.com/user-attachments/assets/85c16e2b-3ba1-447c-a43e-ac22c355ba14)

Scrolling this is still buttery smooth.

I estimate I would need 384 columns  to fill my 3 monitors and maybe 500 columns to fill it without the ASCII view.

Scanning very large data files, like ISO files, allows finding all sorts of interesting patterns, he is a short clip

https://github.com/user-attachments/assets/450682ec-7778-4bcb-8070-bb0aae7766f5

I am curious if it could be also colored for certain patterns, like coloring strings a certain colour, and/or coloring numerical values by their amplitude and recognizing various other data structures like ascii versus unicode   or executable CPU instructions
